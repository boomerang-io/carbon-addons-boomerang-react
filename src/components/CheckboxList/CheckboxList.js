import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'carbon-components-react';
import TooltipHover from '../TooltipHover';
import { Information16 } from '@carbon/icons-react';
import { settings } from 'carbon-components';

const { prefix } = settings;

CheckboxListComponent.propTypes = {
  checkboxProps: PropTypes.object,
  disabled: PropTypes.bool,
  helperText: PropTypes.string,
  id: PropTypes.string,
  initialSelectedItems: PropTypes.array,
  label: PropTypes.string,
  labelText: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.array.isRequired,
  selectedItems: PropTypes.array,
  tooltipClassName: PropTypes.string,
  tooltipContent: PropTypes.any,
  tooltipProps: PropTypes.object,
};

CheckboxListComponent.defaultProps = {
  initialSelectedItems: [],
  tooltipClassName: `${prefix}--bmrg-checkbox-list__tooltip`,
  tooltipProps: { direction: 'top' },
};

function CheckboxListComponent({
  checkboxProps,
  disabled,
  helperText,
  id,
  initialSelectedItems,
  label,
  labelText,
  onChange,
  options,
  selectedItems: propsSelectedItems,
  tooltipClassName,
  tooltipContent,
  tooltipProps,
}) {
  const [stateSelectedItems, setSelectedItems] = useState(initialSelectedItems);

  const selectedItems = propsSelectedItems || stateSelectedItems; // Externally controlled if selectedItems props exists

  const handleCheckboxChange = (value, id, event) => {
    let newSelectedItems = [...stateSelectedItems];

    if (value) {
      newSelectedItems.push(id);
    } else {
      newSelectedItems = newSelectedItems.filter((item) => item !== id);
    }

    setSelectedItems(newSelectedItems);

    if (typeof onChange === 'function') {
      onChange(value, id, event, newSelectedItems);
    }
  };

  const labelValue = label || labelText;

  const labelTextId = !labelValue ? undefined : `${id}-label`;

  return (
    <div key={id} className={`${prefix}--bmrg-checkbox-list`}>
      {labelValue && (
        <div className={`${prefix}--bmrg-checkbox-list__title`}>
          <div id={labelTextId} className={`${prefix}--label`} style={{ marginBottom: '.3125rem' }}>
            {labelValue}
          </div>
          {tooltipContent && (
            <div className={tooltipClassName}>
              <TooltipHover {...tooltipProps} tooltipText={tooltipContent}>
                <Information16 fill="#4d5358" />
              </TooltipHover>
            </div>
          )}
        </div>
      )}
      {helperText && <div className={`${prefix}--form__helper-text`}>{helperText}</div>}

      <div className={`${prefix}--bmrg-checkbox-list__list`}>
        {options.map((option) => {
          const checked = selectedItems.some((item) => item === option.id);
          return (
            <Checkbox
              disabled={disabled}
              key={option.id}
              id={option.id}
              labelText={option.labelText}
              onChange={handleCheckboxChange}
              checked={checked}
              {...checkboxProps}
            />
          );
        })}
      </div>
    </div>
  );
}

export default CheckboxListComponent;
