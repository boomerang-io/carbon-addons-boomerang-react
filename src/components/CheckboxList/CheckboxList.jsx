import React, { useState } from "react";
import PropTypes from "prop-types";
import { Checkbox } from "@carbon/react";
import TooltipHover from "../TooltipHover";
import { Information } from "@carbon/react/icons";
import { prefix } from "../../internal/settings";



CheckboxListComponent.propTypes = {
  checkboxProps: PropTypes.object,
  disabled: PropTypes.bool,
  helperText: PropTypes.string,
  id: PropTypes.string,
  initialSelectedItems: PropTypes.array,
  label: PropTypes.string,
  labelText: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.array.isRequired,
  selectedItems: PropTypes.array,
  tooltipClassName: PropTypes.string,
  tooltipContent: PropTypes.any,
  tooltipProps: PropTypes.object,
};

CheckboxListComponent.defaultProps = {
  initialSelectedItems: [],
  tooltipClassName: `${prefix}--bmrg-checkbox-list__tooltip`,
  tooltipProps: { direction: "top" },
};

function CheckboxListComponent({
  checkboxProps,
  disabled,
  helperText,
  id,
  initialSelectedItems,
  label,
  labelText,
  onChange,
  options,
  selectedItems: propsSelectedItems,
  tooltipClassName,
  tooltipContent,
  tooltipProps,
}) {
  const [stateSelectedItems, setSelectedItems] = useState(initialSelectedItems);

  const selectedItems = propsSelectedItems || stateSelectedItems; // Externally controlled if selectedItems props exists

  const handleCheckboxChange = (event, { checked: value, id }) => {
    let newSelectedItems = [...stateSelectedItems];

    if (value) {
      newSelectedItems.push(id);
    } else {
      newSelectedItems = newSelectedItems.filter((item) => item !== id);
    }

    setSelectedItems(newSelectedItems);

    if (typeof onChange === "function") {
      onChange(value, id, event, newSelectedItems);
    }
  };

  const labelValue = label || labelText;

  const labelTextId = !labelValue ? undefined : `${id}-label`;

  return (
    <div key={id} className={`${prefix}--bmrg-checkbox-list`}>
      {labelValue && (
        <div className={`${prefix}--bmrg-checkbox-list__title`}>
          <div id={labelTextId} className={`${prefix}--label`} style={{ marginBottom: ".3125rem" }}>
            {labelValue}
          </div>
          {tooltipContent && (
            <div className={tooltipClassName}>
              <TooltipHover {...tooltipProps} tooltipText={tooltipContent}>
                <Information size={16} fill="currentColor" />
              </TooltipHover>
            </div>
          )}
        </div>
      )}
      {helperText && <div className={`${prefix}--form__helper-text`}>{helperText}</div>}

      <div className={`${prefix}--bmrg-checkbox-list__list`}>
        {options.map((option) => {
          const checked = selectedItems.some((item) => item === option.id);
          return (
            <Checkbox
              disabled={disabled}
              key={option.id}
              id={option.id}
              labelText={option.labelText}
              onChange={handleCheckboxChange}
              checked={checked}
              {...checkboxProps}
            />
          );
        })}
      </div>
    </div>
  );
}

export default CheckboxListComponent;
