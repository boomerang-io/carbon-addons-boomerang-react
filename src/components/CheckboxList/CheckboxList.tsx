import React, { useState } from "react";
import { Checkbox } from "@carbon/react";
import { Information } from "@carbon/react/icons";
import TooltipHover from "../TooltipHover";
import { prefix } from "../../internal/settings";
import type { TooltipHoverProps } from "../TooltipHover";

type CheckboxListItem = {
  id: string;
  labelText: React.ReactNode;
};

type Props = {
  checkboxProps?: Record<string, any>;
  disabled?: boolean;
  helperText?: string;
  id?: string;
  initialSelectedItems?: string[];
  label?: string;
  labelText?: string;
  onChange?: (
    value: boolean,
    id: string,
    event: React.ChangeEvent<HTMLInputElement>,
    newSelectedItems: string[]
  ) => void;
  options: CheckboxListItem[];
  selectedItems?: string[];
  tooltipClassName?: string;
  tooltipContent?: React.ReactNode;
  tooltipProps?: TooltipHoverProps;
};

function CheckboxListComponent({
  checkboxProps,
  disabled,
  helperText,
  id,
  initialSelectedItems = [],
  label,
  labelText,
  onChange,
  options,
  selectedItems: propsSelectedItems,
  tooltipClassName = `${prefix}--bmrg-checkbox-list__tooltip`,
  tooltipContent,
  tooltipProps = { direction: "top" },
}: Props) {
  const [stateSelectedItems, setSelectedItems] = useState<any[]>(initialSelectedItems);

  const selectedItems = propsSelectedItems || stateSelectedItems; // Externally controlled if selectedItems props exists

  const handleCheckboxChange = (event: any, { checked: value, id }: any) => {
    let newSelectedItems = [...stateSelectedItems];

    if (value) {
      newSelectedItems.push(id);
    } else {
      newSelectedItems = newSelectedItems.filter((item) => item !== id);
    }

    setSelectedItems(newSelectedItems);

    if (typeof onChange === "function") {
      onChange(value, id, event, newSelectedItems);
    }
  };

  const labelValue = label || labelText;

  const labelTextId = !labelValue ? undefined : `${id}-label`;

  return (
    <div key={id} className={`${prefix}--bmrg-checkbox-list`}>
      {labelValue && (
        <div className={`${prefix}--bmrg-checkbox-list__title`}>
          <div id={labelTextId} className={`${prefix}--label`} style={{ marginBottom: ".3125rem" }}>
            {labelValue}
          </div>
          {tooltipContent && (
            <div className={tooltipClassName}>
              <TooltipHover {...tooltipProps} tooltipText={tooltipContent}>
                <Information size={16} fill="currentColor" />
              </TooltipHover>
            </div>
          )}
        </div>
      )}
      {helperText && <div className={`${prefix}--form__helper-text`}>{helperText}</div>}

      <div className={`${prefix}--bmrg-checkbox-list__list`}>
        {options.map((option) => {
          const checked = selectedItems.some((item) => item === option.id);
          return (
            <Checkbox
              disabled={disabled}
              key={option.id}
              id={option.id}
              labelText={option.labelText}
              onChange={handleCheckboxChange}
              checked={checked}
              {...checkboxProps}
            />
          );
        })}
      </div>
    </div>
  );
}

export default CheckboxListComponent;
