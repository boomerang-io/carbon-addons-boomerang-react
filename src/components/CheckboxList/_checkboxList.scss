@include exports('addons-bmrg-checkboxList') {
  .#{$prefix}--bmrg-checkbox-list {
    position: relative;
    width: 100%;
  }

  .#{$prefix}--bmrg-checkbox-list__info {
    align-items: center;
    display: flex;
    position: relative;
  }

  .#{$prefix}--bmrg-checkbox-list__title {
    display: flex;
  }

  .#{$prefix}--bmrg-checkbox-list__tooltip {
    margin-left: 0.5rem;
  }
}
