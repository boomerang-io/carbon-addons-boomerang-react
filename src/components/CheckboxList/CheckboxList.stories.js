import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import CheckboxList from './CheckboxList';

const animals = [
  { labelText: 'Cat', id: 'cat' },
  { labelText: 'Dog', id: 'dog' },
  { labelText: 'Panda', id: 'panda' },
];

const animals2 = [
  { labelText: 'Parrot', id: 'parrot' },
  { labelText: 'Peacock', id: 'peacock' },
  { labelText: 'Penguin', id: 'penguin' },
];

storiesOf('CheckboxList', module)
  .add('default', () => {
    return (
      <CheckboxList
        initialSelectedItems={['panda']}
        onChange={action('checkboxlist changed')}
        options={animals}
      />
    );
  })
  .add('with tooltip and label', () => {
    return (
      <CheckboxList
        helperText="Testing this"
        initialSelectedItems={['peacock']}
        labelText="Select some animals"
        onChange={() => {}}
        options={animals2}
        tooltipContent="Tooltip for checkbox"
        tooltipProps={{ direction: 'top' }}
      />
    );
  });
