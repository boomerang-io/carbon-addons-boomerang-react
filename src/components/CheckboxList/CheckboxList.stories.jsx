import CheckboxList from "./CheckboxList";
import { action } from "@storybook/addon-actions";

export default {
  title: "Inputs/CheckboxList",
  component: CheckboxList,
};

const animals = [
  { labelText: "Cat", id: "cat" },
  { labelText: "Dog", id: "dog" },
  { labelText: "Panda", id: "panda" },
];

const animals2 = [
  { labelText: "Parrot", id: "parrot" },
  { labelText: "Peacock", id: "peacock" },
  { labelText: "Penguin", id: "penguin" },
];

export const Default = () => {
  return <CheckboxList initialSelectedItems={["panda"]} onChange={action("checkboxlist changed")} options={animals} />;
};

export const KitchenSink = () => {
  return (
    <CheckboxList
      helperText="Testing this"
      initialSelectedItems={["peacock"]}
      labelText="Select some animals"
      onChange={() => {}}
      options={animals2}
      tooltipContent="Tooltip for checkbox"
      tooltipProps={{ direction: "top" }}
    />
  );
};
