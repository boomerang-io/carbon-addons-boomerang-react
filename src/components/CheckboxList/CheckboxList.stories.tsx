import CheckboxList from "./CheckboxList";

const animals = [
  { labelText: "Cat", id: "cat" },
  { labelText: "Dog", id: "dog" },
  { labelText: "Panda", id: "panda" },
  { labelText: "Parrot", id: "parrot" },
];

export default {
  title: "Inputs/CheckboxList",
  component: CheckboxList,
  parameters: {
    docs: {
      description: {
        component:
          "Manage checkbox inputs with optional label, helper and tooltip text and use as a data driven component",
      },
    },
  },
  argTypes: {
    options: {
      type: "array",
      defaultValue: animals,
    },
    initialSelectedItems: {
      type: "array",
      defaultValue: ["panda"],
    },
  },
};

export const Default = (args: any) => {
  return <CheckboxList {...args} />;
};

export const KitchenSink = (args: any) => {
  return (
    <CheckboxList
      helperText="Pick a good one"
      initialSelectedItems={["peacock"]}
      labelText="Select some animals"
      onChange={() => {}}
      options={animals}
      tooltipContent="Tooltip for checkbox"
      tooltipProps={{ direction: "top" }}
      {...args}
    />
  );
};
