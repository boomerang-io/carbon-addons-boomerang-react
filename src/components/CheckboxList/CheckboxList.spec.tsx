import { expect, test, vi } from "vitest";
import { render } from "@testing-library/react";

import CheckboxList from "./CheckboxList";

const mockfn = vi.fn();

const animals = [
  { labelText: "Cat", id: "cat" },
  { labelText: "Dog", id: "dog" },
  { labelText: "Panda", id: "panda" },
];

const props = {
  id: "test",
  options: animals,
  helperText: "helper text",
  onChange: mockfn,
  labelText: "label text",
  tooltipContent: "tooltip content",
};

test("render label, helperText, tooltip and options", () => {
  const { queryByText } = render(<CheckboxList {...props} />);
  (expect(queryByText(/helper text/i)) as any).toBeInTheDocument();
  (expect(queryByText(/label text/i)) as any).toBeInTheDocument();
  (expect(queryByText(/cat/i)) as any).toBeInTheDocument();
  (expect(queryByText(/dog/i)) as any).toBeInTheDocument();
  (expect(queryByText(/panda/i)) as any).toBeInTheDocument();
});
