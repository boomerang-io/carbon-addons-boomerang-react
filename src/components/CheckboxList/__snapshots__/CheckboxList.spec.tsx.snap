// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckboxList > snapshot 1`] = `
<body>
  <div>
    <div
      class="cds--bmrg-checkbox-list"
    >
      <div
        class="cds--bmrg-checkbox-list__title"
      >
        <div
          class="cds--label"
          id="test-label"
          style="margin-bottom: .3125rem;"
        >
          label text
        </div>
        <div
          class="cds--bmrg-checkbox-list__tooltip"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.5 11L8.5 6.5 6.5 6.5 6.5 7.5 7.5 7.5 7.5 11 6 11 6 12 10 12 10 11zM8 3.5c-.4 0-.8.3-.8.8S7.6 5 8 5c.4 0 .8-.3.8-.8S8.4 3.5 8 3.5z"
            />
            <path
              d="M8,15c-3.9,0-7-3.1-7-7s3.1-7,7-7s7,3.1,7,7S11.9,15,8,15z M8,2C4.7,2,2,4.7,2,8s2.7,6,6,6s6-2.7,6-6S11.3,2,8,2z"
            />
          </svg>
        </div>
      </div>
      <div
        class="cds--form__helper-text"
      >
        helper text
      </div>
      <div
        class="cds--bmrg-checkbox-list__list"
      >
        <div
          class="cds--form-item cds--checkbox-wrapper"
        >
          <input
            class="cds--checkbox"
            id="cat"
            type="checkbox"
          />
          <label
            class="cds--checkbox-label"
            for="cat"
            title=""
          >
            <span
              class="cds--checkbox-label-text"
              dir="auto"
            >
              Cat
              
            </span>
          </label>
          <div
            class="cds--checkbox__validation-msg"
          />
        </div>
        <div
          class="cds--form-item cds--checkbox-wrapper"
        >
          <input
            class="cds--checkbox"
            id="dog"
            type="checkbox"
          />
          <label
            class="cds--checkbox-label"
            for="dog"
            title=""
          >
            <span
              class="cds--checkbox-label-text"
              dir="auto"
            >
              Dog
              
            </span>
          </label>
          <div
            class="cds--checkbox__validation-msg"
          />
        </div>
        <div
          class="cds--form-item cds--checkbox-wrapper"
        >
          <input
            class="cds--checkbox"
            id="panda"
            type="checkbox"
          />
          <label
            class="cds--checkbox-label"
            for="panda"
            title=""
          >
            <span
              class="cds--checkbox-label-text"
              dir="auto"
            >
              Panda
              
            </span>
          </label>
          <div
            class="cds--checkbox__validation-msg"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
